ucsb-cs156/proj-courses
# pc - fix bugs with quarter selectors

# Summary

In this PR, we do a major refactor of how the START_QTR and END_QTR are configured, and communicated from the backend to the frontend.

We do this in order to simplify the code and make it easier to understand and maintain; the previous version was buggy and difficult to work with.

We also make a few improvements to the Admin / Developer Page to make the contents more clear; it show shows the contents returned by /api/systemInfo, as well as all of the information available through the `env` variable at the time the React build takes place.

# Details

The values of START_QTR and END_QTR are initalized with environment variables (.env on  localhost, and dokku config variables on dokku); this hasn't changed.

What has changed is how those values are communicated to the frontend code.

Previously, those values were picked up by a backend controller, and communicated to the frontend via the /api/systemInfo endpoint. There are several problems with this approach:

* For any information that is communicated from a backend endpoint to a React component, there is a period of time as the component is being first rendered where the values haven't yet been communicated to the component.  This makes it necessary to have logic to deal with how to display the component during the time that the values haven't arrived yet, which complicates the code.
* Furthermore, this delay is unnecessary, since the values are known at application build time.  It would be much more sensible for the frontend code to be able to access those values directly *as if they were constants*.

The new approach defines the variables with the prefix REACT_APP_ which makes them available when the application is being built.

One of the complications is that we want to keep just one `.env` file for environment variables on local host, and we want it to live in the root directory.  However, we note that the environment variable ecosystem for React code (e.g. the `env-cmd` utility) really wants the `.env` file to live in the root of the React code, i.e. in the `frontend` directory, and there doesn't seem to be a way to override this.

Therefore, we need different approaches for the frontend and backend:

# Frontend

* Values are set in `.env` and prefixed with `REACT_APP_` as required by `env-cmd`.
* Before `npm run start` a script is run that copies the values from `.env` at root level (../.env from the perspective of the frontend) into a `.env` file that is a temporary file in the `frontend` directory.

# Backend

* Values are added to Dockerfile as ARG values 
* Values must be declared as build variables via the command 
   ```
   dokku docker-options:add  dockerAppName  build '--build-arg REACT_APP_START_QTR'
   dokku docker-options:add  dockerAppName  build '--build-arg REACT_APP_END_QTR'
   dokku docker-options:add  dockerAppName  build '--build-arg REACT_APP_SOURCE_REPO'
   ```
* Values are picked up during the build phase and are available in `process.env.REACT_APP_START_QTR`, etc.

We have also refactored the QuarterSelector to use the `useLocalStorage` hook so that when values are available in local storage they are used.   This addresses a number of bugs with the quarter selector where it was not behaving as expected (i.e. the value seen on the screen and the value passed to the backend did not correspond.)

# Testing

This branch is currently deployed at https://courses-qa.dokku-00.cs.ucsb.edu

To test, check all of the pages in the app that have quarter selectors.  

* See if they work properly, i.e. if the values selected are the ones that are actually passed to the backend, especially if/when you don't change the value show
* See if they remember their value when you leave the page and return to the page. 
* Login and check the contents of the Admin / Developer Page
